File: node/deliberation/src/specialists/regime.mjs
/**
 * @accretive-substrate/deliberation/specialists/regime
 *
 * Regime specialist. Reads regime_state + relevant regime-preset accretions
 * from the evidence package and emits a verdict based on regime + confidence.
 *
 * Verdict rules (deterministic — no brain calls):
 *   regime=CRASH conf >= 0.8    → hold       (don't add risk into a crash)
 *   regime=BEAR conf >= 0.7     → hold|sell  (sell if open; hold if proposing buy)
 *   regime=BULL conf >= 0.7     → buy        (favor adds + new entries)
 *   regime=SIDEWAYS conf >= 0.7 → mean-reversion bias (buy at lows, sell at highs)
 *                                If trigger doesn't specify direction → hold.
 *   regime=UNKNOWN OR conf<0.5  → abstain (insufficient signal)
 *
 * Cites: regime_state row + any regime-preset accretions in scope.
 *
 * Domain: 'all' — regime applies to every venue.
 */

import { makeSpecialist, voteFor, abstainBecause } from './base.mjs';

export const REGIME_SPECIALIST_NAME = 'regime';

async function argue(pkg) {
  const rs = pkg.regime_state;
  if (!rs || !rs.regime) {
    return abstainBecause(REGIME_SPECIALIST_NAME, 'no regime_state in evidence package');
  }
  const regime = String(rs.regime).toUpperCase();
  const conf = Number(rs.confidence) || 0;

  // Citations: the regime_state row + relevant regime-preset accretions.
  const citations = [
    { type: 'regime_state', regime, confidence: conf, ts: rs.ts || null },
  ];

  // Pull regime-preset accretions matching this regime (already filtered upstream
  // in assemble.mjs, but cite them here for traceability).
  const regimeAccretions = (pkg.accretions || []).filter(a =>
    a.canonical_path === `regime-preset:${regime.toLowerCase().replace(/_/g, '-')}`
  );
  for (const a of regimeAccretions) {
    citations.push({ type: 'accretion', id: a.id, canonical_path: a.canonical_path });
  }

  if (conf < 0.5) {
    return abstainBecause(REGIME_SPECIALIST_NAME, `regime conf ${conf.toFixed(2)} below 0.5 floor`);
  }

  const triggerKind = pkg?.trigger?.kind || 'unknown';

  // CRASH — always hold (don't add risk during freefall).
  if (regime === 'CRASH' && conf >= 0.8) {
    return voteFor(
      REGIME_SPECIALIST_NAME,
      'hold',
      conf,
      `regime=CRASH conf=${conf.toFixed(2)} — refuse new exposure`,
      citations,
    );
  }

  // BEAR — sell if reeval on open position; hold (block) if proposing buy.
  if (regime === 'BEAR' && conf >= 0.7) {
    if (triggerKind === 'sell_reeval') {
      return voteFor(
        REGIME_SPECIALIST_NAME, 'sell', conf,
        `regime=BEAR conf=${conf.toFixed(2)} — close on re-eval`, citations,
      );
    }
    return voteFor(
      REGIME_SPECIALIST_NAME, 'hold', Math.min(0.85, conf + 0.1),
      `regime=BEAR conf=${conf.toFixed(2)} — block new buys, no add`, citations,
    );
  }

  // BULL — favor buys.
  if (regime === 'BULL' && conf >= 0.7) {
    if (triggerKind === 'buy_proposal' || triggerKind === 'stack_add') {
      return voteFor(
        REGIME_SPECIALIST_NAME, 'buy', conf,
        `regime=BULL conf=${conf.toFixed(2)} — favor entry`, citations,
      );
    }
    return voteFor(
      REGIME_SPECIALIST_NAME, 'hold', conf * 0.8,
      `regime=BULL conf=${conf.toFixed(2)} but trigger=${triggerKind} not direction-aligned`, citations,
    );
  }

  // SIDEWAYS — mean-reversion bias. Without further per-symbol context here,
  // give a modest-confidence buy on buy_proposal (favor mean-reversion long entries)
  // and a modest-confidence sell on sell_reeval (take profit when stretched).
  if (regime === 'SIDEWAYS' && conf >= 0.7) {
    if (triggerKind === 'buy_proposal' || triggerKind === 'stack_add') {
      return voteFor(
        REGIME_SPECIALIST_NAME, 'buy', conf * 0.85,
        `regime=SIDEWAYS conf=${conf.toFixed(2)} — mean-reversion long bias`, citations,
      );
    }
    if (triggerKind === 'sell_reeval') {
      return voteFor(
        REGIME_SPECIALIST_NAME, 'sell', conf * 0.7,
        `regime=SIDEWAYS conf=${conf.toFixed(2)} — take profit on stretch`, citations,
      );
    }
    return voteFor(
      REGIME_SPECIALIST_NAME, 'hold', 0.5,
      `regime=SIDEWAYS conf=${conf.toFixed(2)} trigger=${triggerKind} ambiguous`, citations,
    );
  }

  // Below thresholds.
  return abstainBecause(
    REGIME_SPECIALIST_NAME,
    `regime=${regime} conf=${conf.toFixed(2)} below decision thresholds`,
  );
}

export const regimeSpecialist = makeSpecialist(REGIME_SPECIALIST_NAME, ['all'], argue);

File: go/deliberation/pkg/deliberation/specialists/regime/regime.go
// Package regime — Go port of @accretive-substrate/deliberation/specialists/regime.
//
// Verdict logic MUST match the Node side byte-for-byte for the cross-
// pipeline parity claim. Fixture tests at
// <this repo>/go/deliberation/internal/test/fixtures/ are loaded by BOTH
// Node and Go suites with identical input/output assertions.
package regime

import (
	"fmt"
	"strings"
	"time"

	"github.com/ramene/accretive-substrate/go/deliberation/pkg/deliberation"
)

const Name = "regime"

func Argue(pkg deliberation.EvidencePackage) deliberation.Voice {
	rs := pkg.RegimeState
	if rs == nil || rs.Regime == "" {
		return deliberation.AbstainBecause(Name, "no regime_state in evidence package")
	}
	regime := strings.ToUpper(rs.Regime)
	conf := rs.Confidence

	// Cite regime_state + matching regime-preset accretions.
	citations := []deliberation.Citation{{
		Type:       "regime_state",
		Regime:     regime,
		Confidence: conf,
		TS:         rs.TS,
	}}

	// Match accretions: canonical_path = regime-preset:<regime-lowercased-with-dashes>
	target := "regime-preset:" + strings.ReplaceAll(strings.ToLower(regime), "_", "-")
	for _, a := range pkg.Accretions {
		if a.CanonicalPath == target {
			citations = append(citations, deliberation.Citation{
				Type:          "accretion",
				ID:            a.ID,
				CanonicalPath: a.CanonicalPath,
			})
		}
	}

	if conf < 0.5 {
		return deliberation.AbstainBecause(Name, fmt.Sprintf("regime conf %.2f below 0.5 floor", conf))
	}

	triggerKind := string(pkg.Trigger.Kind)
	if triggerKind == "" {
		triggerKind = "unknown"
	}

	// CRASH — always hold.
	if regime == "CRASH" && conf >= 0.8 {
		return deliberation.VoteFor(Name, deliberation.VerdictHold, conf,
			fmt.Sprintf("regime=CRASH conf=%.2f — refuse new exposure", conf), citations)
	}

	// BEAR — sell on sell_reeval, hold on buy_proposal.
	if regime == "BEAR" && conf >= 0.7 {
		if triggerKind == "sell_reeval" {
			return deliberation.VoteFor(Name, deliberation.VerdictSell, conf,
				fmt.Sprintf("regime=BEAR conf=%.2f — close on re-eval", conf), citations)
		}
		boosted := conf + 0.1
		if boosted > 0.85 {
			boosted = 0.85
		}
		return deliberation.VoteFor(Name, deliberation.VerdictHold, boosted,
			fmt.Sprintf("regime=BEAR conf=%.2f — block new buys, no add", conf), citations)
	}

	// BULL — favor buy.
	if regime == "BULL" && conf >= 0.7 {
		if triggerKind == "buy_proposal" || triggerKind == "stack_add" {
			return deliberation.VoteFor(Name, deliberation.VerdictBuy, conf,
				fmt.Sprintf("regime=BULL conf=%.2f — favor entry", conf), citations)
		}
		return deliberation.VoteFor(Name, deliberation.VerdictHold, conf*0.8,
			fmt.Sprintf("regime=BULL conf=%.2f but trigger=%s not direction-aligned", conf, triggerKind), citations)
	}

	// SIDEWAYS — mean-reversion bias.
	if regime == "SIDEWAYS" && conf >= 0.7 {
		if triggerKind == "buy_proposal" || triggerKind == "stack_add" {
			return deliberation.VoteFor(Name, deliberation.VerdictBuy, conf*0.85,
				fmt.Sprintf("regime=SIDEWAYS conf=%.2f — mean-reversion long bias", conf), citations)
		}
		if triggerKind == "sell_reeval" {
			return deliberation.VoteFor(Name, deliberation.VerdictSell, conf*0.7,
				fmt.Sprintf("regime=SIDEWAYS conf=%.2f — take profit on stretch", conf), citations)
		}
		return deliberation.VoteFor(Name, deliberation.VerdictHold, 0.5,
			fmt.Sprintf("regime=SIDEWAYS conf=%.2f trigger=%s ambiguous", conf, triggerKind), citations)
	}

	return deliberation.AbstainBecause(Name,
		fmt.Sprintf("regime=%s conf=%.2f below decision thresholds", regime, conf))
}

// Specialist — exported safe-wrapped argue function.
var Specialist = deliberation.MakeSafe(Name, Argue)

// Avoid unused-import warning when time is only used in citation struct.
var _ = time.Time{}
